tests/system/test_dbapi.py

```diff
@@ -544,6 +570,245 @@ def _insert_row(self, i):
+        method_count_interceptor.reset()
+        # called 3 times
+        self._insert_row(1)
+        dbapi_database._method_abort_interceptor.set_method_to_abort(
+            EXECUTE_STREAMING_SQL_METHOD
+        )
+        # called 3 times
+        self._cursor.execute("SELECT * FROM contacts")
+        dbapi_database._method_abort_interceptor.reset()
+        self._cursor.fetchall()
+        # called 2 times
+        self._insert_row(2)
+        # called 2 times
+        self._cursor.execute("SELECT * FROM contacts")
+        self._cursor.fetchone()
+        dbapi_database._method_abort_interceptor.set_method_to_abort(COMMIT_METHOD)
+        # called 2 times
+        self._conn.commit()
+        dbapi_database._method_abort_interceptor.reset()
+        assert method_count_interceptor._counts[COMMIT_METHOD] == 2
+        assert method_count_interceptor._counts[EXECUTE_STREAMING_SQL_METHOD] == 10
+
+        self._cursor.execute("SELECT * FROM contacts")
+        got_rows = self._cursor.fetchall()
+        assert len(got_rows) == 2
+
+    def test_consecutive_aborted_transactions(self, dbapi_database):
+        """Test 2 consecutive transactions with Abort exceptions on the same
+        connection works."""
+
+        method_count_interceptor = dbapi_database._method_count_interceptor
+        method_count_interceptor.reset()
+        self._insert_row(1)
+        self._insert_row(2)
+        self._cursor.execute("SELECT * FROM contacts")
+        self._cursor.fetchall()
+        dbapi_database._method_abort_interceptor.set_method_to_abort(COMMIT_METHOD)
+        self._conn.commit()
+        dbapi_database._method_abort_interceptor.reset()
+        assert method_count_interceptor._counts[COMMIT_METHOD] == 2
+        assert method_count_interceptor._counts[EXECUTE_STREAMING_SQL_METHOD] == 6
+
+        method_count_interceptor = dbapi_database._method_count_interceptor
+        method_count_interceptor.reset()
+        self._insert_row(3)
+        self._insert_row(4)
+        self._cursor.execute("SELECT * FROM contacts")
+        self._cursor.fetchall()
+        dbapi_database._method_abort_interceptor.set_method_to_abort(COMMIT_METHOD)
+        self._conn.commit()
+        dbapi_database._method_abort_interceptor.reset()
+        assert method_count_interceptor._counts[COMMIT_METHOD] == 2
+        assert method_count_interceptor._counts[EXECUTE_STREAMING_SQL_METHOD] == 6
+
+        self._cursor.execute("SELECT * FROM contacts")
+        got_rows = self._cursor.fetchall()
+        assert len(got_rows) == 4
+
+    @pytest.mark.noautofixt
+    def test_abort_retry_multiple_cursors(self, shared_instance, dbapi_database):
+        """Test that retry works when multiple cursors are involved in the transaction."""
+
+        try:
+            conn = Connection(shared_instance, dbapi_database)
+            cur = conn.cursor()
+            cur.execute(
+                """
+                CREATE TABLE Singers (
```

Comment:
nit: Creating an extra table just for this test is relatively expensive. Could we instead just use one table, and instead separate the cursors by letting cursor1 select the rows with id (1, 2), and cursor2 select the rows with id (3 4)?

Reply:
Thanks, done